src/cider/nrepl/middleware/debug.clj

```diff
@@ -28,18 +28,22 @@
   "Send value and coordinates to the client through the debug channel.
   Sends a response to the message stored in debugger-message."
   [value extras]
-  ;; Notify cider that a (break) is incoming, along with the value of
-  ;; the instrumented sexp and instructions on how to find it in the
-  ;; code.
-  (transport/send
-   (:transport @debugger-message)
-   (response-for @debugger-message
-                 (assoc extras
-                        :debug-value (pr-str value)
-                        :breakfunction nil)))
-  ;; Send the actual break.
-  (binding [*out* (new java.io.StringWriter)]
-    (break value)))
+  ;; Don't debug code being run synchronously. It is probably
+  ;; unintended and will unexpectedly ask for input.
+  (if (:sync clojure.tools.nrepl.middleware.interruptible-eval/*msg*)
```

Comment:
> Not been following the issue too closely, so forgive my ignorance, but shouldn't this really be fixed client-side (i.e. properly handle need-input during synchronous requests) so we don't need this kind of hack?

Yes, this should be fixed client-side, but I'm under the impression that even with the handle-input fix in sync-request the breakpoint will still bug out during a sync eval (because its workings are more complicated than a simple need-input event). Still, I haven't had a chance to test that conjecture yet (busy busy busy :-P).